### ISSUE TITLE
gitstatus as non-daemon

### ISSUE THREAD

**Issue by billziss-gh** [romkatv/gitstatus#71]
(Apologies if this has been discussed before, but a quick search through the issues did not immediately reveal the answer to me.)

Your `gitstatus` project is particularly intriguing. Unfortunately `gitstatus` runs as a daemon and in general I avoid running daemons unless they provide some major system functionality.

With that in mind I am wondering if it would be possible to run `gitstatus` as a standalone program that does not daemonize itself. I understand that the advantage of a daemon is that you can store state between uses and perhaps watch for file system changes. However perhaps `gitstatus` could still be fast and useful even without that daemonized state.

I also note that at least on Windows and macOS there are facilities to watch for file system changes even if the software is not running:

- Windows: [Change Journals](https://docs.microsoft.com/en-us/windows/win32/fileio/change-journals)
- macOS: [File System Events](https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/FSEvents_ProgGuide/UsingtheFSEventsFramework/UsingtheFSEventsFramework.html)

**Comment by romkatv** [comment]
> Your `gitstatus` project is particularly intriguing. Unfortunately `gitstatus` runs as a daemon and in general I avoid running daemons unless they provide some major system functionality.
> With that in mind I am wondering if it would be possible to run `gitstatus` as a standalone program that does not daemonize itself.

gitstatusd (the native binary) doesn't daemonize itself. From [How it works](https://github.com/romkatv/gitstatus/blob/master/README.md#how-it-works):

> gitstatusd reads requests from stdin and prints responses to stdout. Requests contain an ID and a directory. Responses contain the same ID and machine-readable git status for the directory.
>
> [ZSH bindings](https://github.com/romkatv/gitstatus/blob/master/gitstatus.plugin.zsh) and [Bash bindings](https://github.com/romkatv/gitstatus/blob/master/gitstatus.plugin.sh) start gitstatusd in the background and communicate with it via pipes.

From `gitstatusd --help`:

```text
EXAMPLE

  Send a single request and print response (zsh syntax):

    local req_id=id
    local dir=$PWD
    echo -nE $req_id$'\x1f'$dir$'\x1e' | ./gitstatusd | {
      local resp
      IFS=$'\x1f' read -rd $'\x1e' -A resp && print -lr -- "${(@qq)resp}"
    }

  Output:
    'id'
    '1'
    '/home/romka/gitstatus'
    'bf46bf03dbab7108801b53f8a720caee8464c9c3'
    'master'
    'master'
    'origin'
    'git@github.com:romkatv/gitstatus.git'
    ''
    '70'
    '1'
    '0'
    '0'
    '2'
    '0'
    '0'
    ''
    '0'
```

> I understand that the advantage of a daemon is that you can store state between uses ...

Correct. gitstatusd uses a different diffing algorithm from regular `git`. This algorithm requires Git index in a different format. Since gitstatusd has no control over the format of Git index written to disk by `git`, it needs to transform it. This takes time. The converted index can be used many times, so it pays off to keep it around.

There are a few other things that gitstatusd needs to compute on the first request but doesn't have to repeat afterwards.

> ... and perhaps watch for file system changes.

gitstatusd doesn't watch for file system changes. See [this comment](https://github.com/romkatv/gitstatus/commit/050aaaa04b652e15ab4c28b669f1ed753f9d0519#commitcomment-33077677) and the discussion below for the reasoning behind this decision.

> However perhaps `gitstatus` could still be fast and useful even without that daemonized state.

It takes `max(1s, 10 * T)` to start gitstatusd, send it a request, receive a response and shut it down. `T` is the time to process the same request when gitstatusd is daemonized and 10 is approximate.

The `1s` lower bound can be removed without much difficulty, which will reduce latency to `10 * T`. This is roughly the same latency as plain `git status --porcelain=v2 -b`, so there is little reason to use gitstatusd at all.

These numbers are from [Benchmarks](https://github.com/romkatv/gitstatus/blob/master/README.md#benchmarks):

> | Tool          |      Cold  |         Hot |
> |---------------|-----------:|------------:|
> | **gitstatus** | **291 ms** | **30.9 ms** |
> | git           |     876 ms |      295 ms |
> | lg2           |    1730 ms |     1310 ms |

Note that gitstatus/cold is roughly equal to git/hot and 10 times slower than gitstatus/hot.

> I also note that at least on Windows and macOS there are facilities to watch for file system changes even if the software is not running:

As I mentioned above, gitstatusd doesn't watch for file system changes.

**Comment by billziss-gh** [comment]
> gitstatusd (the native binary) doesn't daemonize itself.

Thank you for taking the time to explain this to me.

However I still have some confusion:

- Your comment "T is the time to process the same request when gitstatusd is **daemonized**" suggests that `gitstatusd` is somehow daemonized.
- Your comment "ZSH bindings and Bash bindings start gitstatusd in the **background** and communicate with it via pipes" suggests that `gitstatusd` is somehow placed in the background (perhaps as a background shell job).

To resolve this confusion of mine I ended up finding [`gitstatus_start`](https://github.com/romkatv/powerlevel10k/blob/master/gitstatus/gitstatus.plugin.zsh#L280) where I see that `gitstatusd` is indeed "daemonized" by a [double invocation of zsh](https://github.com/romkatv/powerlevel10k/blob/master/gitstatus/gitstatus.plugin.zsh#L395-L399). So `gitstatusd` does not daemonize itself, but is "daemonized" by your shell code.

> gitstatusd doesn't watch for file system changes.

Understood.

> The 1s lower bound can be removed without much difficulty, which will reduce latency to 10 * T. This is roughly the same latency as plain git status --porcelain=v2 -b, so there is little reason to use gitstatusd at all

Understood.

> gitstatusd uses a different diffing algorithm from regular git. This algorithm requires Git index in a different format. Since gitstatusd has no control over the format of Git index written to disk by git, it needs to transform it. This takes time. The converted index can be used many times, so it pays off to keep it around.

Let me rephrase this to make sure I understand: It sounds like your special sauce is that `gitstatus` needs its data in a format different from git's native format. You run `gitstatus` in the background in order to maintain that data as state in memory (and not to watch for file system changes).

I am wondering then if an alternative might be to store the state as a file in the `.git` directory. This file would be re-read on every invocation of `gitstatus` to restore the previous state and then proceed with the rest of the `gitstatus` logic.

While this scheme might be slightly slower than the current "daemon" scheme, I find it has some distinct advantages:

- Avoids the need for the "daemon" / "background" process.
- Is likely easier to integrate with foreign prompts (like mine -- I do not use themes).
- Is likely easier to port to platforms like Windows.

Please note that I am not asking you to do this work (although it would be great if you did since you have a good understanding of thee problem). I am just exploring this as a possibility.

**Comment by romkatv** [comment]
> So `gitstatusd` does not daemonize itself, but is "daemonized" by your shell code.

Correct. gitstatusd doesn't daemonize itself but it can be daemonized externally. The official shell bindings do exactly this.

> Let me rephrase this to make sure I understand: It sounds like your special sauce is that `gitstatus` needs its data in a format different from git's native format. You run `gitstatus` in the background in order to maintain that data as state in memory (and not to watch for file system changes).

This is one of the reasons why "hot" requests complete faster than "cold" requests in gitstatusd. There are others. The long-lived state of gitstatusd informed dozens of design decisions and optimizations, so the hot and cold codepaths have many differences.

> I am wondering then if an alternative might be to store the state as a file in the `.git` directory. This file would be re-read on every invocation of `gitstatus` to restore the previous state and then proceed with the rest of the `gitstatus` logic.

It's certainly feasible to dump and restore some of the state that gitstatusd creates on the first cold request.

> While this scheme might be slightly slower

For small repositories it'll likely be an order of magnitude slower, although it's not that important (except on windows, where forks are prohibitively expensive). For large repositories it might be possible to be just 3-5x slower but I'm not sure.

> than the current "daemon" scheme, I find it has some distinct advantages:
> 
> * Avoids the need for the "daemon" / "background" process.

Why is it this an advantage? What is the disadvantage of having a background process? Genuinely curious.

> * Is likely easier to integrate with foreign prompts (like mine -- I do not use themes).

The words "theme" and "prompt" are used interchangeably, at least in zsh world. Technically, theme is code that sets prompt.

```zsh
# Zsh theme called "minimalist".
function prompt_minimalist_setup() { PROMPT='%m%# ' }
```

If you mean that it's more difficult to integrate gitstatusd with your prompt rather than fetching data from `git`, I don't think this is true. Here's an example from the docs that defines theme/prompt with Git status on the right:

```zsh
source ~/gitstatus/gitstatus.plugin.zsh

function my_set_prompt() {
  PROMPT='%~%# '
  RPROMPT=''

  if gitstatus_query MY && [[ $VCS_STATUS_RESULT == ok-sync ]]; then
    RPROMPT=${${VCS_STATUS_LOCAL_BRANCH:-@${VCS_STATUS_COMMIT}}//\%/%%}  # escape %
    (( $VCS_STATUS_NUM_STAGED    )) && RPROMPT+='+'
    (( $VCS_STATUS_NUM_UNSTAGED  )) && RPROMPT+='!'
    (( $VCS_STATUS_NUM_UNTRACKED )) && RPROMPT+='?'
  fi

  setopt noprompt{bang,subst} promptpercent  # enable/disable correct prompt expansions
}

gitstatus_stop 'MY' && gitstatus_start -s -1 -u -1 -c -1 -d -1 'MY'
autoload -Uz add-zsh-hook
add-zsh-hook precmd my_set_prompt
```

Achieving the same result (ignoring performance) with plain `git` will require more code and it'll be more complex.

> * Is likely easier to port to platforms like Windows.

How so? gitstatus currently works on Windows via WSL and via Cygwin. It doesn't work on mingw because I've never attempt to compile gitstatusd for it.

> Please note that I am not asking you to do this work (although it would be great if you did since you have a good understanding of thee problem). I am just exploring this as a possibility.

Understood. I'm currently not very motivated to explore this direction because it's difficult for me to see the benefits while the downsides are obvious (slower serving, more complex code). Making gitstatusd easier to use (e.g., by making it daemonize itself automatically and then talk to the daemon), or making it faster, or giving it a more generic API -- these are the things I find valuable.

**Comment by billziss-gh** [comment]
> For small repositories it'll likely be an order of magnitude slower, although it's not that important (except on windows, where forks are prohibitively expensive). For large repositories it might be possible to be just 3-5x slower but I'm not sure.

I use the standard `git-prompt.sh`, which I believe can spawn multiple git processes in order to do its work. Having a single `gitstatus` process report all necessary information would still be a win for me even if it was an order of magniture slower that `gitstatusd` (which would put it at 300ms according to your benchmarks).

> Why is it this an advantage? What is the disadvantage of having a background process? Genuinely curious.

Running any program is a benefit vs risk proposition. I am running a program to get some benefit, but I am also accepting a certain risk. For example, I am running my C compiler to compile my program, but I also run the (minimal) risk that it will crash my computer, trash my files or steal my credit card details.

For this reason a transient program that provides me a benefit is preferable to a "background" program that provides me the same benefit (assuming that both programs are of the same quality). The "background" program has simply more opportunities to introduce a dangerous behavior.

Background programs may increase memory pressure, increase CPU time and decrease battery time, introduce security holes, etc.

I should clarify that I am not saying that `gitstatusd` is doing any of these things. I am only explaining some of the reasons that I avoid running "background" processes when I have the option.

> How so? gitstatus currently works on Windows via WSL and via Cygwin. It doesn't work on mingw because I've never attempt to compile gitstatusd for it.

Understood. I often use Cygwin when on Windows. But I also use Powershell where you can modify the prompt.

>  I'm currently not very motivated to explore this direction because it's difficult for me to see the benefits while the downsides are obvious (slower serving, more complex code).

Perfectly understandable. Feel free to close the issue if you want.

**Comment by billziss-gh** [comment]
> If you are OK with an order of magnitude slowdown, perhaps `git status --porcelain=2 -b` will serve your needs?

Thanks for the suggestion. If this provides me all the information that I get from `git-prompt.sh` I may use it.

> I'm closing the issue as we seem to have understood each other sufficiently and no one is committing to any immediate action.

Understood. Thanks for considering my request.
